## Machine Learning Record Mining

Project to create a pipeline that uses GeoDeepDive's output to find Unaquired Sites for Neotoma.

Using NLP parsed text and a Data Science approach, identify whether a paper is suitable for Neotoma and detect features such as 'Site Name', 'Location', 'Age Span' and 'Site Descriptions'.

In [33]:
# Loading libraries

import numpy as np
import pandas as pd
import csv
import psycopg2
import re
import ast

In [18]:
# Options for DF display
pd.set_option('display.max_colwidth', 50)
pd.set_option('display.max_rows', 10)

#pd.set_option('display.max_rows', None)
#pd.set_option('display.max_columns', None)
#pd.set_option('display.width', None)
#pd.set_option('display.max_colwidth', -1)

## Loading and viewing the Data

### Loading NLP Sentences

In [28]:
# Connect to PostgreSQL server from terminal:
# pg_ctl -D PSQL_Data -l logfile start
# Data obtained after running data_loader.py

nlp_sentences = pd.read_csv('src/output/nlp_sentences_df.csv', delimiter='\t')

In [29]:
nlp_sentences.head(2)

_gddid  sentid  \
0  54b43266e138239d8684efed       1   
1  54b43266e138239d8684efed       2   

                                             wordidx  \
0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   
1            [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]   

                                               words  \
0  ['Available', 'online', 'at', 'www.sciencedire...   
1  ['The', 'Chihuahueños', 'Bog', 'record', 'exte...   

                                      part_of_speech  \
0  ['JJ', 'NN', 'IN', 'NNP', 'NNP', 'NNP', 'CD', ...   
1  ['DT', 'NNP', 'NN', 'NN', 'VBZ', 'TO', 'IN', '...   

                                       special_class  \
0  ['O', 'O', 'O', 'O', 'O', 'O', 'NUMBER', 'O', ...   
1  ['O', 'O', 'O', 'O', 'O', 'O', 'O', 'NUMBER', ...   

                                              lemmas  \
0  ['available', 'online', 'at', 'www.sciencedire...   
1  ['the', 'Chihuahueños', 'bog', 'record', 'exte...   

                                           word_type  \
0  ['dep', 'dep', 'dep', 'dep', 'dep', 'dep', 'de...   
1  ['det', 'compound', 'compound', 'nsubj', '', '...   

                                       word_modified  \
0  [218, 218, 218, 218, 218, 218, 218, 0, 218, 0,...   
1          [4, 4, 4, 5, 0, 11, 11, 11, 11, 11, 5, 0]   

                                     words_as_string dms_regex dd_regex  
0  Available,online,at,www.sciencedirect.com,Quat...        []       []  
1  The,Chihuahueños,Bog,record,extends,to,over,15...        []       []

In [5]:
# Running this takes forever. Do not run unless necessary
# nlp_sentences.describe(include = 'all')

In [30]:
avg_sentences = nlp_sentences.groupby('_gddid')['_gddid','sentid'].transform('max')
avg_sentences.drop_duplicates(subset='_gddid', keep='first', inplace=True)
avg_sentences.describe()

sentid
count   218.000000
mean    513.600917
std     232.460891
min       2.000000
25%     397.250000
50%     504.000000
75%     638.250000
max    1330.000000

From the Summary Statistics, despite having 111,965 sentences, there are only 218 unique articles (`_gddid`). The article with most sentences, had 1330 sentences; that's the max number found in `sentid`.  
In order to know the average of sentences, I grouped by `_gddid` and then, obtained the number of sentences per article. After that, I just got the mean through a Summary of statistics to know that on average, each article has 514 sentences.

### Loading Bibliography Data

In [23]:
bibliography = pd.read_csv('src/output/bibliography.csv', delimiter=',')

In [24]:
bibliography.head(3)

_type                            _id                    publisher  \
0   doi  10.1016/S0277-3791(99)00007-4                     Elsevier   
1   doi                10.1139/b92-002  Canadian Science Publishing   
2   doi   10.1016/0031-0182(80)90043-7                     Elsevier   

                                               title  \
0  Palaeoclimate, chronology and vegetation histo...   
1  Holocene history of forest trees on the Bruce ...   
2  Glacial sequence and environmental history in ...   

                                   journal.name.name  \
0                         Quaternary Science Reviews   
1                         Canadian Journal of Botany   
2  Palaeogeography, Palaeoclimatology, Palaeoecology   

                                              author  year number  volumne  \
0  Huntley, B.; Watts, W.A.; Allen, J.R.M.; Zolit...  1999      7      NaN   
1                                     Bennett, K. D.  1992      1      NaN   
2  Van Der Hammen, T.; Barelds, J.; De Jong, H.; ...  1980    NaN      NaN   

                     _gddid     type     pages  \
0  550453fde1382326932d85f7  article  945--960   
1  578b5aabcf58f1587003efba  article     6--18   
2  54b43244e138239d868493cd  article  247--340   

                                            link_url  link_type  
0  http://www.sciencedirect.com/science/article/p...  publisher  
1  http://www.nrcresearchpress.com/doi/abs/10.113...  publisher  
2  http://www.sciencedirect.com/science/article/p...  publisher

## EDA

Reviewing our data includes skimming through some papers online and seeing if the data is consistent with our NLP Sentences dataframe.

From there, we can also visualize what we would like our model to predict: 'Location', 'Site Name', 'Age Span', and 'Site Description' from a "Human perspective".

### Finding Latitude and Longitud via REGEX 

After reviewing some PDF's and viewing the data parsing, I noticed some possible patterns for coordinates.

<img src="figures/img/00_snippet_of_pdf.png" alt="Drawing" style="width: 300px;"/>

Which translates to our pandas dataframe as:

![alt text](figures/img/01_snippet_in_sql.png "PandasVersion")

As a first problem, we can notice that the degrees symbol `°` was translated as a `3`. On some other occassions, it is confused as an `o` or a middle point `◦`. For `'`, this symbol got confused by the word `cents`.

This repo will use normal REGEX to find coordinates in the most intuitive way. 

A coordinate for longitude is a number ranging from -180° to 180°. It has three components, `°, ',"` and it covers the position in E or W.

Latitude goes from -90° to 90°, has the same three components and goes from S to N.

The data_loader.py uses the 2 standard `dms` and `dd` readings and found the coordinates using REGEX.

In [26]:
#nlp_sentences[['words_as_string', 'words', 'dms_regex']].describe()

The `dms_regex` only returned 95 unique values. The `dd_regex` has only 8 unique values. That's to say that from the 111965 sentences, only 103 possess something similar to the REGEX we are looking for.  

Two examples of REGEX:

In [31]:
# dms regex example
dms_example = nlp_sentences[nlp_sentences['_gddid'] == '550453fde1382326932d85f7']
dms_example = dms_example[dms_example['sentid'] == 10]
dms_example[['words', 'words_as_string', 'dms_regex', 'dd_regex']]

words  \
94875  ['Introduction', 'Laghi', 'di', 'Monticchio', ...   

                                         words_as_string  \
94875  Introduction,Laghi,di,Monticchio,are,two,maar,...   

                           dms_regex dd_regex  
94875  ['403,56,40,N,,,153,36,48,E']       []

In [32]:
# dd_regex example
dd_example = nlp_sentences[nlp_sentences['_gddid'] == '57a12aaccf58f1b0962a836c']
dd_example = dd_example[dd_example['sentid'] == 62]
dd_example[['words', 'words_as_string', 'dms_regex', 'dd_regex']]

words  \
62439  ['Kingston', 'meadow', 'Kingston', 'Meadow', '...   

                                         words_as_string dms_regex  \
62439  Kingston,meadow,Kingston,Meadow,-LRB-,2400,m,e...        []   

                   dd_regex  
62439  ['117.11,W,39.21,N']

### Problems with the REGEX approach

As seen before, extracting information using REGEX is very complicated. We would require a lot of queries that would retrieve each particular case.

A possible alternative could be using NER, Named Entity Recognition.

In [14]:
def order_article(article_id):
    '''
    Function to find an article by its gddid in the NLP sentences and have it displayed in order
    
    Keyword arguments:
    article_id -- gddid
    
    Returns:
    article ordered by sentence index
    '''
    article = nlp_sentences[nlp_sentences['_gddid'] == article_id]
    return article[['sentid', 'words']].sort_values(by = 'sentid')

Article 01  
Skimmed info: 
 * `Article Name:` Palaeoclimate_chronology_and_vegetation_history_of_the_Weichselian_Lateglacial
 * `Link`: http://www.sciencedirect.com/science/article/pii/S0277379199000074  
 * `Site Name`:  Laghi di Monticchio
 * `Location`:  40° 56' 40" N, 15° 36' 48" E
 * `Age Span`: 
 * `Site Descriptions`: The record from Lago Grande di Monticchio now adds macrofossil evidence of the presence of relatively boreal tree taxa during the Lateglacial interstadial and provides a chronology based upon annually laminated sediments.

In [12]:
# Article 01 
order_article('550453fde1382326932d85f7')

sentid                                              words
94865       1  [Quaternary, Science, Reviews, 18, -LRB-, 1999...
94866       2  [Plant, macrofossil, data, from, one, core, pr...
94867       3  [Quantitative, palaeoclimate, reconstructions,...
94868       4  [The, chronological, framework, for, Lateglaci...
94870       5  [Surface, conditions, in, the, North, Atlantic...
...       ...                                                ...
95325     448  [Vegetation, history, and, climate, of, the, l...
95326     449  [Quaternary, Science, Reviews, ,, 15, ,, 133, ...
95327     450  [Zolitschka, ,, B., ,, &, Negendank, ,, J., F....
95329     451  [Sedimentology, ,, dating, and, palaeoclimatic...
95330     452  [Quaternary, Science, Reviews, ,, 15, ,, 101, ...

[452 rows x 2 columns]

Article 02  
Skimmed info:  
* `Article:` Development of the Mixed Connifer Forest in Northern New Mexico and its relationship to Holocene environmental change
* `Link`: http://www.sciencedirect.com/science/article/pii/S0033589407001512
* `Site Name`:  Chihuahuenos Bog 
* `Location`:  ??? No Given Coordinates 
* `Age Span`:   
* `Site Descriptions`:  Chihuahueños Bog is located within the mixed conifer forest and has affinities to the flora of the southern Rocky Mountains(Foxx and Tierney, 1985; Allen, 2004). Although the area immediately surrounding the bog has been heavily logged in recent years, nearly every conifer species growing in montaneforests of the Jemez mountains is represented either along theshore of the bog or in the surrounding uplands.  

In [34]:
# Article 2
# order_article('54b43266e138239d8684efed')

Article 03  
Skimmed info:  
* `Link`: http://dx.doi.org/10.1130/g35541.1  # No Full access to article
* `Site Name`:  Cahokia region, Illinois, USA
* `Location`:  N38°42'00" - N38°42'00", W90°04'60" - W90°04'60" - Not found on text (yet).
* `Age Span`:   
* `Site Descriptions`: Record from Horseshoe Lake, an oxbow lake in the central Mississippi River valley that is adjacent to the Cahokia site (Illinois, USA), the largest prehistoric settlement north of Mexico.   

In [35]:
# Article 3
# order_article('57c5b941cf58f1338eaddb5b')

Article 04  
Skimmed info:  
* `Link:`http://www.tandfonline.com/doi/abs/10.1080/00173130902965157
* `Site Name`:  Begbunar
* `Location`:  42°09′ N, 22° 33′ E; 1750 ; in SQL file shows as: `42,°,09,cents,N,,,22,°,33,cents,E`
* `Age Span`:   
* `Site Descriptions`: Located in the central treeless zone of the Osogovo mountains, which are situated at the border between south-western Bulgaria and the north-eastern former Yugoslav Republic of Macedonia.

In [37]:
# Article 4
# order_article('58d29193cf58f14928755ba5')

In [38]:
# Try to see the exact location. Wont show last few words.
# pd.DataFrame(order_article('58d29193cf58f14928755ba5').iloc[4])

Article 05  
Skimmed info:  
* `Link`: http://dx.doi.org/10.1130/0016-7606(1997)109<1306:tfalqp>2.3.co
* `Site Name`:  Cottonwood III, Cottonwood IV, Red Lady, Red Well, Splains
* `Location`:  38°49'50' 106°24'45', 38°49'50' 106°24'45', 38°52'50' 107°2'30', 38°52'50' 107°2'30',38°53'40' 107°3'15', 38°52'28' 107°05', 38°50' 107°4'30', 38°50' 107°4'30', 38°52' 107°2'30', 38°45' 106°50', 38°45' 106°50'
* `Age Span`:   17000 - 400
* `Site Descriptions`:   The lower timberline or lower forest border is the lower elevational limit of a continuous forest belt. In the Rocky Mountains it forms a sharp contrast with shrublands or grasslands in valley bottoms. Trees may extend below the lower timberline in favorable habitats, along streams or on rocky outcrops.  

In [39]:
# Article 5
# order_article('57928e07cf58f133d1c26609')

In [40]:
# Print the exact line where locations are.
# pd.DataFrame(order_article('57928e07cf58f133d1c26609').iloc[179])

In [42]:
# Find the article in the bibliography by gddid to retrieve Article Name

bibliography[bibliography['_gddid'] == '57928e07cf58f133d1c26609']

_type                                              _id publisher  \
93   doi  10.1130/0016-7606(1997)109<1306:tfalqp>2.3.co;2       GSA   

                                                title  \
93  Timberline fluctuations and late Quaternary pa...   

                         journal.name.name             author  year number  \
93  Geological Society of America Bulletin  Fall, Patricia L.  1997     10   

    volumne                    _gddid     type       pages  \
93      NaN  57928e07cf58f133d1c26609  article  1306--1320   

                                             link_url  link_type  
93  http://dx.doi.org/10.1130/0016-7606(1997)109<1...  publisher

## Linking Neotoma DB to extract locations

In [43]:
neotoma = pd.read_csv("../Do_not_commit_data/data-1590729612420.csv")
neotoma = neotoma[['siteid', 'sitename', 'longitudeeast', 'latitudenorth', 'longitudewest', 'latitudesouth', 'sitedescription', 'doi']]

neotoma.head(3)

siteid       sitename  longitudeeast  latitudenorth  longitudewest  \
0   10330  Lac du Sommet      -70.66468       47.71662      -70.66573   
1   10330  Lac du Sommet      -70.66468       47.71662      -70.66573   
2    1729    Myrtle Lake      -93.37853       47.98645      -93.39207   

   latitudesouth                                    sitedescription  \
0       47.71382  The small shallow Lac du Sommet (0.02 km2, 4 m...   
1       47.71382  The small shallow Lac du Sommet (0.02 km2, 4 m...   
2       47.97876  Lake surrounded by peatland. Physiography: Red...   

                        doi  
0  10.1177/0959683611400199  
1  10.1177/0959683611400199  
2           10.1139/b68-190

In [44]:
nlp_bib = nlp_sentences.merge(bibliography, on = '_gddid')
nlp_bib = nlp_bib.rename(columns={'_id': 'doi'})
nlp_bib.head(3)

_gddid  sentid  \
0  54b43266e138239d8684efed       1   
1  54b43266e138239d8684efed       2   
2  54b43266e138239d8684efed       3   

                                             wordidx  \
0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   
1            [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]   
2  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   

                                               words  \
0  ['Available', 'online', 'at', 'www.sciencedire...   
1  ['The', 'Chihuahueños', 'Bog', 'record', 'exte...   
2  ['An', 'Artemisia', 'steppe', ',', 'then', 'an...   

                                      part_of_speech  \
0  ['JJ', 'NN', 'IN', 'NNP', 'NNP', 'NNP', 'CD', ...   
1  ['DT', 'NNP', 'NN', 'NN', 'VBZ', 'TO', 'IN', '...   
2  ['DT', 'NNP', 'NN', ',', 'RB', 'DT', 'JJ', 'NN...   

                                       special_class  \
0  ['O', 'O', 'O', 'O', 'O', 'O', 'NUMBER', 'O', ...   
1  ['O', 'O', 'O', 'O', 'O', 'O', 'O', 'NUMBER', ...   
2  ['O', 'O', 'O', 'O', 'O', 'O', 'O', 'LOCATION'...   

                                              lemmas  \
0  ['available', 'online', 'at', 'www.sciencedire...   
1  ['the', 'Chihuahueños', 'bog', 'record', 'exte...   
2  ['a', 'Artemisia', 'steppe', ',', 'then', 'a',...   

                                           word_type  \
0  ['dep', 'dep', 'dep', 'dep', 'dep', 'dep', 'de...   
1  ['det', 'compound', 'compound', 'nsubj', '', '...   
2  ['det', 'compound', 'nsubj', '', 'advmod', 'de...   

                                       word_modified  \
0  [218, 218, 218, 218, 218, 218, 218, 0, 218, 0,...   
1          [4, 4, 4, 5, 0, 11, 11, 11, 11, 11, 5, 0]   
2  [3, 3, 10, 0, 9, 9, 9, 9, 10, 0, 14, 14, 14, 1...   

                                     words_as_string  ...  \
0  Available,online,at,www.sciencedirect.com,Quat...  ...   
1  The,Chihuahueños,Bog,record,extends,to,over,15...  ...   
2  An,Artemisia,steppe,,,then,an,open,Picea,woodl...  ...   

                                               title    journal.name.name  \
0  Development of the mixed conifer forest in nor...  Quaternary Research   
1  Development of the mixed conifer forest in nor...  Quaternary Research   
2  Development of the mixed conifer forest in nor...  Quaternary Research   

                                              author  year number volumne  \
0  Anderson, R. Scott; Jass, Renata B.; Toney, Ja...  2008      2     NaN   
1  Anderson, R. Scott; Jass, Renata B.; Toney, Ja...  2008      2     NaN   
2  Anderson, R. Scott; Jass, Renata B.; Toney, Ja...  2008      2     NaN   

      type     pages                                           link_url  \
0  article  263--275  http://www.sciencedirect.com/science/article/p...   
1  article  263--275  http://www.sciencedirect.com/science/article/p...   
2  article  263--275  http://www.sciencedirect.com/science/article/p...   

   link_type  
0  publisher  
1  publisher  
2  publisher  

[3 rows x 25 columns]

In [46]:
print(nlp_bib.shape)

(111965, 25)


In [48]:
nlp_bib[['words_as_string', 'journal.name.name', 'dms_regex', 'dd_regex']].head(3)

words_as_string    journal.name.name  \
0  Available,online,at,www.sciencedirect.com,Quat...  Quaternary Research   
1  The,Chihuahueños,Bog,record,extends,to,over,15...  Quaternary Research   
2  An,Artemisia,steppe,,,then,an,open,Picea,woodl...  Quaternary Research   

  dms_regex dd_regex  
0        []       []  
1        []       []  
2        []       []

In [49]:
by_doi = neotoma.groupby('doi')
neotoma_sitenames = by_doi.apply(lambda x: list(x['sitename']))
neotoma_siteid = by_doi.apply(lambda x: list(x['siteid']))
neotoma_longeast = by_doi.apply(lambda x: list(x['longitudeeast']))
neotoma_latnorth = by_doi.apply(lambda x: list(x['latitudenorth']))

neotoma_joined_df = neotoma_sitenames.to_frame('sitenames').join(neotoma_siteid.to_frame('siteid')).join(neotoma_longeast.to_frame('longeast')).join(neotoma_latnorth.to_frame('latnorth'))
neotoma_joined_df = neotoma_joined_df.reset_index()

neotoma_joined_df.head(5)

doi  \
0                                     0.1139/e80-122   
1  10.1002/(SICI)1099-1417(199607/08)11:4<327::AI...   
2  10.1002/(SICI)1099-1417(199701/02)12:1<15::AID...   
3  10.1002/(SICI)1099-1417(199801/02)13:1<43::AID...   
4  10.1002/1099-1417(200010)15:7<687::AID-JQS551>...   

                                     sitenames              siteid  \
0  [East Baltic Bog, Mermaid Bog, Portage Bog]   [763, 1684, 1903]   
1                               [Lundin Tower]              [3256]   
2      [Spruce Pond, Spruce Pond, Spruce Pond]  [2508, 2508, 2508]   
3     [Pantano de Genagra, Pantano de Genagra]      [14030, 14030]   
4       [Laguna Loma Linda, Laguna Loma Linda]      [13971, 13971]   

                            longeast  \
0         [-62.15, -63.025, -64.075]   
1                         [-2.97206]   
2  [-74.18236, -74.18236, -74.18236]   
3             [-76.60428, -76.60428]   
4             [-73.35661, -73.35661]   

                                            latnorth  
0                      [46.408333, 46.25, 46.673611]  
1                                         [56.21529]  
2  [41.237770000000005, 41.237770000000005, 41.23...  
3                                   [2.4751, 2.4751]  
4                                 [3.30805, 3.30805]

From Neotoma DB, we have obtained a dataframe grouped by DOI that lists each `sitename`, `siteid`, `longitude east`, and `latitunde north`. Now, let's join this information to the NLP_Bib dataframe. For now, let's do to the one that coordinates have been obtained with REGEX just to see if some of the numbers and sitenames make sense. 

In [51]:
nlp_bib_neotoma = nlp_bib.merge(neotoma_joined_df, on = 'doi', how = 'left')
print(nlp_bib_neotoma.shape)

(111965, 29)


In [52]:
# Filter df to just frames that has REGEX longer than one
mask = (nlp_bib_neotoma['dms_regex'].str.len() > 0)
nlp_bib_neotoma_regex = nlp_bib_neotoma.loc[mask]
nlp_bib_neotoma_regex[['words','dms_regex']].head(100)

print(nlp_bib_neotoma_regex.shape)

(111965, 29)


In [195]:
nlp_bib_neotoma_regex = nlp_bib_neotoma_regex[['_gddid', 'sentid', 'doi', 'words', 'words_as_string', 'dms_regex', 'sitenames', 'latnorth', 'longeast']]
nlp_bib_neotoma_regex = nlp_bib_neotoma_regex.reset_index()
nlp_bib_neotoma_regex.head(1)

index                    _gddid  sentid              doi  \
0   3314  578a1f6ecf58f1b54b810cd9      94  10.1139/b85-070   

                                               words  \
0  [Lake, Hope, Simpson, -LRB-, 52, ', 27, ', N, ...   

                                     words_as_string  \
0  Lake,Hope,Simpson,-LRB-,52,',27,',N,,,56,',26,...   

                     dms_regex                               sitenames  \
0  [52,',27,',N,,,56,',26,',W]  [Lake Hope Simpson, 18 [Moraine Lake]]   

             latnorth              longeast  
0  [52.45, 52.266667]  [-56.433333, -58.05]

We can see that most of the REGEX indeed pulled out the same locations for latitude and longitudes as in the Neotoma database.

In [196]:
nlp_bib_neotoma[['_gddid', 'sentid', 'doi', 'words', 'dms_regex', 'sitenames', 'latnorth', 'longeast']].head(2)

_gddid  sentid                          doi  \
0  54b43266e138239d8684efed       1  10.1016/j.yqres.2007.12.002   
1  54b43266e138239d8684efed       2  10.1016/j.yqres.2007.12.002   

                                               words dms_regex  \
0  [Available, online, at, www.sciencedirect.com,...        []   
1  [The, Chihuahueños, Bog, record, extends, to, ...        []   

                              sitenames            latnorth  \
0  [Chihuahueños Bog, Chihuahueños Bog]  [36.0491, 36.0491]   
1  [Chihuahueños Bog, Chihuahueños Bog]  [36.0491, 36.0491]   

                   longeast  
0  [-106.50963, -106.50963]  
1  [-106.50963, -106.50963]

In [244]:
nlp_bib_neotoma['latnorth_s'] = nlp_bib_neotoma['latnorth'].apply(lambda x: [int(i) for i in x])
nlp_bib_neotoma['latnorth_s'] = nlp_bib_neotoma['latnorth_s'].apply(lambda x: [str(i) for i in x])

nlp_bib_neotoma['longeast_s'] = nlp_bib_neotoma['longeast'].apply(lambda x: [int(i) for i in x])
nlp_bib_neotoma['longeast_s'] = nlp_bib_neotoma['longeast_s'].apply(lambda x: [str(i) for i in x])

In [250]:
# nlp_bib_neotoma[['latnorth_s', 'longeast_s']].head(-2)

In [252]:
nlp_bib_neotoma['intersection_words_lat'] = nlp_bib_neotoma[['words','latnorth_s']].apply(lambda x : list(set.intersection(*map(set,list(x)))),axis=1)
nlp_bib_neotoma['intersection_words_long'] = nlp_bib_neotoma[['words','longeast_s']].apply(lambda x : list(set.intersection(*map(set,list(x)))),axis=1)
df = nlp_bib_neotoma[(nlp_bib_neotoma['intersection_words_lat'].apply(len) > 0) & (nlp_bib_neotoma['intersection_words_long'].apply(len) > 0 )]
df2 = nlp_bib_neotoma[nlp_bib_neotoma['intersection_words_long'].apply(len) > 0 ]
print(df2.shape)
df2[['words', 'longeast_s', 'latnorth_s', 'intersection_words_lat', 'intersection_words_long']]



(219, 32)


words  \
5261    [Holarctic, Ecology, 2, ,, 65, --, 74, .]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             
8446    [Site, details, The, alder-carr, fen, Na, Bahne, -LRB-, 240, m, a., s., l., ,, 50, °, 11, ', 51.4, '', N, ,, 15, °, 57, ', 32.9, '', E, ,, 1, ha, -RRB-, lies, below, the, youngest, ,, Würmian, terrace, in, a, floodplain, ,, 150, m, from, the, Orlice, River, ,, about, 8, km, east, of, the, Elbe, River, .]                                                                                                                                                                                                                                                                                                                                                                                                                                                                     
9273    [Table, 2, Radiocarbon, age, measurements, Depth, -LRB-, cm, -RRB-, Material, Dry, weight, Laboratory, ;, -LRB-, mg, -RRB-, sample, number, 28, --, 29, 41, --, 42, 63, --, 64, 77, --, 78, 116, --, 117, 122, --, 123, 127, --, 128, Betula, twig, ,, woody, fragments, and, Empetrum, leaves, Betula, twig, ,, woody, fragments, and, Empetrum, leaves, Woody, fragments, ,, Empetrum, leaves, and, leaf, fragments, Woody, fragments, and, Empetrum, leaves, Twig, ,, woody, fragments, ,, Empetrum, and, Juniperus, leaves, ,, leaf, fragments, Empetrum, leaves, and, plant, debris, Woody, fragments, ,, Empetrum, leaves, and, leaf, fragments, 4.5, 3.7, 6.7, 15.5, 13.2, 6.3, ...]                                                                                           
9790    [Computers, &, Geosciences, 28, ,, 421, --, 426, .]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
11223   [The, sediments, contain, well-preserved, remains, of, plants, ,, diatoms, and, chironomids, and, indicate, :, -LRB-, 1, -RRB-, a, mosaic, pioneer, vegetation, of, dwarf-shrub, tundra, ,, snow, bed, vegetation, and, xerophytes, on, wind-blown, ridges, ,, -LRB-, 2, -RRB-, an, icefree, lake, during, summers, with, clear, water, at, low, pH, due, to, little, inwash, ,, -LRB-, 3, -RRB-, a, continental, climate, with, summers, slightly, warmer, than, today, -LRB-, ca, 10, 1C, -RRB-, ,, and, most, probably, with, very, cold, winters, ,, according, to, indications, of, permafrost, ,, and, -LRB-, 4, -RRB-, a, hiatus/reduced, sedimentation, during, at, least, the, first, half, of, the, ...]                                                                     
11261   [This, tree-line, roughly, follows, the, 10, 1C, July, T, isotherm, in, continental, areas, -LRB-, Odland, ,, 1996, -RRB-, .]                       

In [253]:
df3 = nlp_bib_neotoma[nlp_bib_neotoma['intersection_words_lat'].apply(len) > 0 ]
print(df3.shape)
df3[['words', 'longeast_s', 'latnorth_s', 'intersection_words_lat', 'intersection_words_long']]

(634, 32)


words  \
393     [Quaternary, Research, 36, ,, 307, --, 321, .]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       
681     [INDEX, LAT, -LRB-, ◦, N, -RRB-, LONG, -LRB-, ◦, W, -RRB-, ha, m, mg, μg, /, L, μS, /, cm, CaCo3/L, mg/L, mg/L, mg/L, mg/L, mg/L, mg/L, μg, /, L, Acidification, 1, Chocolate, 44, ◦, 38.3, 63, ◦, 37.3, 7.1, 13.0, 6.2, 2, 497.0, 1.5, 11.6, 76.7, 122.0, 26.4, 0.3, 0.4, 0.1, 2, First, Chain, 44, ◦, 38.3, 63, ◦, 38.3, 72.6, 13.9, 4.8, 2, 328.0, −, 1.1, 5.5, 48.1, 80.3, 18.3, 0.1, 0.9, 0.4, 3, Frenchman, 44, ◦, 41.8, 63, ◦, 34.5, 8.4, 1.5, 6.1, 12, 469.0, 0.6, 4, Major, 44, ◦, 43.7, 63, ◦, 28.9, 343.0, 65.0, 4.7, ...]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 
1998    [Ret, ', iew, of, Palaeobotany, und, Palynology, 44, ,, 277-292, .]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  
2154    [2, ,, T., 114N, ,, R., 26W, -LRB-, 44, '', 45, ', N., lat., ,, 93, '', 55, ', W., long, ., -RRB-]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   

## Intersections Between Sitenames and NLP Sentences
Let's now extract Sitenames from the Neotoma database. For this, I will group by doi and then, list all sitenames that are mentioned in each doi.

In [63]:
nlp_bib_neotoma['words_l']  = nlp_bib_neotoma['words'].astype(str).str.lower().transform(ast.literal_eval)


In [64]:
nlp_bib_neotoma['sitenames_l'] = nlp_bib_neotoma['sitenames'].astype(str).str.lower().transform(ast.literal_eval)
nlp_bib_neotoma[['words_l', 'sitenames_l']].head(2)

words_l  \
0  [available, online, at, www.sciencedirect.com,...   
1  [the, chihuahueños, bog, record, extends, to, ...   

                            sitenames_l  
0  [chihuahueños bog, chihuahueños bog]  
1  [chihuahueños bog, chihuahueños bog]

In [65]:
nlp_bib_neotoma['sitenames_l']

0           [chihuahueños bog, chihuahueños bog]
1           [chihuahueños bog, chihuahueños bog]
2           [chihuahueños bog, chihuahueños bog]
3           [chihuahueños bog, chihuahueños bog]
4           [chihuahueños bog, chihuahueños bog]
                           ...                  
106635    [peat-bog begbunar, peat-bog begbunar]
106636    [peat-bog begbunar, peat-bog begbunar]
106637    [peat-bog begbunar, peat-bog begbunar]
106638    [peat-bog begbunar, peat-bog begbunar]
106639    [peat-bog begbunar, peat-bog begbunar]
Name: sitenames_l, Length: 106640, dtype: object

In [102]:
nlp_bib_neotoma['intersection_words_sitenames'] = nlp_bib_neotoma[['words_l','sitenames_l']].apply(lambda x : list(set.intersection(*map(set,list(x)))),axis=1)
#nlp_bib_neotoma['intersection_words_sitenames']

df = nlp_bib_neotoma[nlp_bib_neotoma['intersection_words_sitenames'].apply(len) > 0 ]
df[['words_l', 'sitenames_l', 'intersection_words_sitenames', 'dms_regex']]


words_l  \
681     [index, lat, -lrb-, ◦, n, -rrb-, long, -lrb-, ...   
788     [seven, lakes, -lrb-, bayers, ,, bell, ,, cran...   
2110    [quaternary, research, 16, ,, 66-79, -lrb-, 19...   
2111    [and, h., e., wright, ,, jr., ~, *, department...   
2120    [the, geologic, and, ecologic, succession, at,...   
...                                                   ...   
94912   [m, a, p, of, central, europe, showing, the, s...   
95048   [distribution, of, mollusc, assemblages, -lrb-...   
95138   [1983, -rrb-, ,, the, krumpa, a, l, l, e, r, ~...   
106480  [kožlí, -lrb-, s., bohemia, ,, czech, republic...   
106481  [kožlí, -lrb-, s., bohemia, ,, czech, republic...   

                                              sitenames_l  \
681     [anderson lake, anderson lake, bayers, bayers,...   
788     [anderson lake, anderson lake, bayers, bayers,...   
2110                          [norwood, norwood, norwood]   
2111                          [norwood, norwood, norwood]   
2120                          [norwood, norwood, norwood]   
...                                                   ...   
94912                            [krumpa, krumpa, krumpa]   
95048                            [krumpa, krumpa, krumpa]   
95138                            [krumpa, krumpa, krumpa]   
106480                                     [kožlí, kožlí]   
106481                                     [kožlí, kožlí]   

       intersection_words_sitenames dms_regex  
681                        [bayers]        []  
788                        [bayers]        []  
2110                      [norwood]        []  
2111                      [norwood]        []  
2120                      [norwood]        []  
...                             ...       ...  
94912                      [krumpa]        []  
95048                      [krumpa]        []  
95138                      [krumpa]        []  
106480                      [kožlí]        []  
106481                      [kožlí]        []  

[438 rows x 4 columns]

In [56]:
print(type(nlp_bib_neotoma['sitenames'][48]))
print(type(nlp_bib_neotoma['words'][48]))
print(type(nlp_bib_neotoma['words_as_string'][48]))

<class 'list'>
<class 'list'>
<class 'str'>
